src/composables/secureStorageRef.ts

```diff
@@ -0,0 +1,137 @@
+    if (encryptionData) {
+      try {
+        const decryptedValue = await decrypt(encryptionData, encryptedValue);
+        try {
+          return JSON.parse(decryptedValue);
+        } catch (e) {
+          return decryptedValue;
+        }
+      } catch (e) {
+        throw new Error('Failed to decrypt the value');
+      }
+    }
+    throw new Error('Failed to read the value, probably the password is not set');
+  }
+
+  async function getEncryptedValue(val: any): Promise<T> {
+    if (IS_MOBILE_APP || !val) {
+      return val;
+    }
+    const { getEncryptionData } = useAccounts();
+    const encryptionData = await getEncryptionData();
+    if (encryptionData) {
+      return encrypt(encryptionData, val) as T;
+    }
+    throw new Error('Failed to write the value');
+  }
+
+  const decryptedState = ref(initialState) as Ref<T>;
+
+  /** Should always hold encrypted values (both on storage and state) */
+  const innerState = useStorageRef<T>(null as T, storageKey, {
+    ...options,
+    enableSecureStorage: true,
+    // Handle write operation within the composable but read from storageRef
+    serializer: {
+      read: getDecryptedValue,
+      write: (val) => val,
+    },
+    onRestored: async (val: T) => {
+      if (val) {
+        const decryptedValue = await getDecryptedValue(val);
+        decryptedState.value = decryptedValue;
+        options.onRestored?.(decryptedValue);
+      }
+    },
+    onBackgroundSync: async (val: T) => {
+      decryptedState.value = await getDecryptedValue(val);
+    },
+  });
+
+  /**
+   * This is an exception for the mnemonic because it's a key part of the password feature.
+   * It restores the mnemonic from the insecure storage and sets it on the accounts composable.
+   */
+  async function setMnemonicOnAccounts(mnemonic: T) {
+    if (mnemonic && validateMnemonic(mnemonic)) {
+      const { setMnemonicAndInitializePassword } = useAccounts();
+      await setMnemonicAndInitializePassword(mnemonic as string, true);
+      options.onRestored?.(mnemonic);
+    }
+  }
+
+  // Restore state and run watchers
+  (async () => {
+    // Check if value exists in insecure storage and migrate it to secure storage
+    const decryptedValue = await WalletStorage.get<T | null>(storageKey);
+    if (decryptedValue && storageKey === STORAGE_KEYS.mnemonic) {
+      await setMnemonicOnAccounts(decryptedValue);
+    }
+
+    if (IS_MOBILE_APP) {
+      // TODO Needs to be tested
+      // Move the unencrypted value to secure storage and remove it from insecure storage
+      if (decryptedValue !== null) {
+        decryptedState.value = decryptedValue;
+        WalletStorage.remove(storageKey);
+      }
+    }
+
+    /**
+     * Clear the state when the user logs out.
+     * This ensures that the state is not leaked if someone removes the modal from the DOM.
+     */
+    const { encryptionData } = useAccounts();
+    watch(encryptionData, async (val) => {
```

Comment:
is it correct that val is only used in the if case here instead of reading from it?

Reply:
`encryptionData` has been updated and is now `encryptionKey` so `val` would be the `key`. We don't want to read from the `key` but we want to set/reset the `decryptedState` based on whether we have a `key` or not